application/controllers/collection_instrument.py

```diff
@@ -276,48 +272,34 @@ def upload_eq(self, survey_id, classifiers=None, session=None):
+    def update_exercise_eq_instruments(self, exercise_id: str, instruments: list, session=None) -> bool:
         """
-        Both linking and unlinking of collection instruments is carried out here.
-
-        :param instrument_selection: A list of instrument ids
-        :param exercise_id: A collection exercise id (UUID)
+        Updates eQ instruments for an exercise. Current instruments are used to determine which ones should be
+        appended and/or removed.
+        :param exercise_id: The collection exercise id
+        :param instruments: A list of instruments that the collection exercise should now have
         :param session: database session
+        :return: a bool
         """
-        validate_uuid(exercise_id)
-
-        linked_instruments = []
-
-        # This query will either create a CE (with no CIs attached) or will pull in all linked CIs for the exercise id
-        # In question.
-        exercise_and_instruments = self._find_or_create_exercise(exercise_id, session)
-
-        for i in exercise_and_instruments.instruments:
-            linked_instruments.append(str(i.instrument_id))
 
-        instruments_to_add = set(instrument_selection).difference(linked_instruments)
-        instruments_to_remove = set(linked_instruments).difference(instrument_selection)
-
-        if instruments_to_add:
-            self.update_collection_exercise_with_cis(instruments_to_add, True, exercise_and_instruments, session)
+        validate_uuid(exercise_id)
+        exercise = self._find_or_create_exercise(exercise_id, session)
+        current_instruments = [str(instrument.instrument_id) for instrument in exercise.instruments]
 
-        if instruments_to_remove:
-            self.update_collection_exercise_with_cis(instruments_to_remove, False, exercise_and_instruments, session)
+        instruments_to_add = set(instruments).difference(current_instruments)
+        instruments_to_remove = set(current_instruments).difference(instruments)
 
-    def update_collection_exercise_with_cis(self, instruments, to_add, exercise_and_instruments, session):
-        for instrument_id in instruments:
-            validate_uuid(instrument_id)
+        for instrument_id in instruments_to_add:
+            instrument = self.get_instrument_by_id(instrument_id, session)
+            instrument.exercises.append(exercise)
 
+        for instrument_id in instruments_to_remove:
             instrument = self.get_instrument_by_id(instrument_id, session)
+            instrument.exercises.remove(exercise)
 
-            if to_add:
-                instrument.exercises.append(exercise_and_instruments)
-            else:
-                instrument.exercises.remove(exercise_and_instruments)
+        log.info("Collection instruments updated successfully", instruments=instruments, exercise_id=exercise_id)
 
-        self.publish_collection_instrument_to_collection_exercise(
```

Comment:
The cause of the bug, being in the same db_session (transaction) the publish step will effectively behaviour before any of the append remove happens, as they won't persist till the whole of the transaction has passed